File: knapsack.c
#include<stdio.h>
int number;
int knapsack;
int remainingcapacity;
struct weighted{
	float weight;
	float profit;
	float ratio;
	int object;
}quantity[10];
void readArray(struct weighted array[10])
{
   printf("\nEnter Weights\n");
	for(int i=0;i<number;i++)
	{
		printf("Enter weight of %d :  ",(i+1));
		scanf("%f,",&array[i].weight);
		array[i].object=i+1;
	}
	printf("\nEnter Values\n");
	for(int i=0;i<number;i++)
	{
		printf("Enter value of %d :  ",(i+1));
		scanf("%f,",&array[i].profit);
	}
	

}
void displayArray(struct weighted array[10])
{
	printf("\nObject\tWeight\tValue\tRatio\n");
	for(int i=0;i<number;i++)
	{
		printf("  %d\t  %.0f\t  %.0f\t %f\n",array[i].object,array[i].weight,array[i].profit,array[i].ratio);
	}
}

void sort(struct weighted array[10])
{
    for(int i=0;i<number;i++)
    {
        for(int j=0;j<i;j++)
        {
            if(array[i].ratio>array[j].ratio)
            {
                struct weighted temp=array[i];
				array[i]=array[j];
				array[j]=temp;
                
             
            }
        }
    }
	
}

void greedyKnapsack(struct weighted array[10])
{
	int i;
	float items_in_knapsack[10];
	float profit=0;
	printf("\nItems in knapsack: ");
	for( i=0;i<number;i++)
	{
		items_in_knapsack[i]=0;
	}
		remainingcapacity=knapsack;
		for(i=0;i<number;i++)
		{
			if(array[i].weight>remainingcapacity){
			break;
			}
			printf("%d ",array[i].object);
		     items_in_knapsack[i]=1;
			remainingcapacity-=array[i].weight;
			profit=profit+array[i].profit;
		}
		if(i<number){
		items_in_knapsack[i]=remainingcapacity/array[i].weight;
		printf("%d ",array[i].object);
		profit=profit+items_in_knapsack[i]*(array[i].profit);
		printf("(%.2f)\n",items_in_knapsack[i]);
		}
		printf("The profit is:%f",profit);
	
}
	


void main()
{
	
	printf("Enter knapsack capacity:");
	scanf("%d",&knapsack);
	printf("enter number of objects:");
	scanf("%d",&number);
	readArray(quantity);
	printf("\nObject\tWeight\tValue\n");
	for(int i=0;i<number;i++)
	{
		printf("  %d\t  %.0f\t  %.0f\n",quantity[i].object,quantity[i].weight,quantity[i].profit);
	}
	printf("\ncalculating ratio\n");
	for(int i=0;i<number;i++)
	{
		quantity[i].ratio=quantity[i].profit/quantity[i].weight;
	}
	displayArray(quantity);
	printf("\nSorting\n");
	sort(quantity);
	displayArray(quantity);
	greedyKnapsack(quantity);
    
}

File: Knapsack.java
import java.util.Scanner;

public class Knapsack {

    public static int knapSack(int W, int wt[], int val[], int n) {
        int i, w;
        int K[][] = new int[n + 1][W + 1];

        for (i = 0; i <= n; i++) {
            for (w = 0; w <= W; w++) {
                if (i == 0 || w == 0)
                    K[i][w] = 0;
                else if (wt[i - 1] <= w)
                    K[i][w] = Math.max(val[i - 1] + K[i - 1][w - wt[i - 1]], K[i - 1][w]);
                else
                    K[i][w] = K[i - 1][w];
            }
        }
        
        return K[n][W];
    }


    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of items: ");
        int n = sc.nextInt();

        int val[] = new int[n];
        int wt[] = new int[n];

        System.out.println("Enter the values of the items:");
        for (int i = 0; i < n; i++) {
            val[i] = sc.nextInt();
        }

        System.out.println("Enter the weights of the items:");
        for (int i = 0; i < n; i++) {
            wt[i] = sc.nextInt();
        }

        System.out.print("Enter the maximum weight capacity: ");
        int W = sc.nextInt();

        int maxVal = knapSack(W, wt, val, n);
        System.out.println("Maximum value that can be obtained = " + maxVal);
    }
}